### ISSUE TITLE
After updating the GIT repository, the following steps in Weblate are not performed

### ISSUE THREAD

**Issue by SlavekB** [WeblateOrg/weblate#2813]
**Describe the bug**
I run my own Weblate instance linked to repositories managed by Gitea. In Gitea, a webhook is set to update components in Weblate when commit to Gitea. The add-on gettext.msgmerge is used in Weblate to update PO files after commit with update of POT files. Usually everything works perfectly smoothly and excellently.

Sometimes, however, after commit to Gitea, a rebase repository is executed in Weblate, but then no other steps are taken at all – no component updates, no msgmerge for updated POT files, therefore no updates to PO files, no error message in the logs.

**To Reproduce**
What I saw as common in such situations, here were translations waiting for commit in some of components linked to the same git repository.

**Expected behavior**
I expect the repository update to be performed whenever an update is invoked in Weblate and new commits are fetched from git.

**Server configuration and status**
```
 * Weblate 3.6.1
 * Python 3.5.3
 * Django 1.11.20
 * Celery 4.2.2
 * celery-batches 0.2
 * six 1.10.0
 * social-auth-core 3.1.0
 * social-auth-app-django 3.1.0
 * django-appconf 1.0.1
 * translate-toolkit 2.3.1
 * translation-finder 1.1
 * Whoosh 2.7.0
 * defusedxml 0.5.0
 * Git 2.11.0
 * Pillow 4.0.0
 * python-dateutil 2.5.3
 * lxml 4.2.5
 * django-crispy-forms 1.6.1
 * django_compressor 2.2
 * djangorestframework 3.9.2
 * user-agents 1.1.0
 * jellyfish 0.7.1
 * diff-match-patch 20121119
 * pytz 2016.7
 * pyuca 1.1.2
 * PyYAML 3.12
 * Database backends: django.db.backends.mysql
 * Cache backends: default:RedisCache, avatar:FileBasedCache
 * Celery: redis://localhost:6379, redis://localhost:6379, regular
 * Platform: Linux 3.16.6-042stab134.46 (x86_64)
```
```
System check identified some issues:

INFOS:
?: (weblate.I021) Error collection is not configured, it is highly recommended for production use
        HINT: https://docs.weblate.org/en/weblate-3.6.1/admin/install.html#collecting-errors

System check identified 1 issue (0 silenced).
```

**Additional context**
In the log "committing pending changes" is mentioned, but nothing else follows:
```
[2019-06-12 21:11:46,537: INFO/ForkPoolWorker-5] Lock 140331270848920 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-06-12 21:11:46,538: INFO/ForkPoolWorker-5] Lock 140331270848920 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-06-12 21:11:46,762: INFO/ForkPoolWorker-5] tde/tde-i18n: updating repository
[2019-06-12 21:11:46,762: INFO/ForkPoolWorker-5] Lock 140331270848920 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-06-12 21:11:49,898: INFO/ForkPoolWorker-5] tde/tde-i18n: update took 3.14 seconds
[2019-06-12 21:11:51,097: INFO/ForkPoolWorker-5] Lock 140331270848920 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-06-12 21:11:55,217: INFO/ForkPoolWorker-5] Lock 140331270848920 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-06-12 21:11:56,419: INFO/ForkPoolWorker-5] tde/tde-i18n: rebase remote into repo
[2019-06-12 21:12:02,472: INFO/ForkPoolWorker-5] tdeutils/ktimer/tr: committing pending changes (addon)
[2019-06-12 21:14:02,526: INFO/ForkPoolWorker-5] Lock 140331270848920 released on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
```

**Comment by nijel** [comment]
The confusing commit pending might be caused by e4e5abcdac351db9031b720d893d212d2d763934. Besides that the log looks okay, looks like there were no changes on the files. Maybe check other worker logs and look for errors.

**Comment by SlavekB** [comment]
Great – it looks like an exact fix. I did backporting the mentioned commit to 3.6.1. We'll see what happens next time in a similar situation.

Thank you!

**Comment by SlavekB** [comment]
Looks like nothing has changed after applying the patch.

The situation was the same – the only pending update of translation, the POT file update was pushed (another linked component in the same repository), and nothing else happened after the repository was updated:
```
[2019-06-19 22:11:48,989: INFO/ForkPoolWorker-2] Lock 139657724594272 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-06-19 22:11:49,010: INFO/ForkPoolWorker-2] Lock 139657724594272 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-06-19 22:11:49,103: INFO/ForkPoolWorker-2] tde/tde-i18n: updating repository
[2019-06-19 22:11:49,104: INFO/ForkPoolWorker-2] Lock 139657724594272 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-06-19 22:11:57,104: INFO/ForkPoolWorker-2] tde/tde-i18n: update took 8.00 seconds
[2019-06-19 22:11:58,301: INFO/ForkPoolWorker-2] Lock 139657724594272 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-06-19 22:12:03,545: INFO/ForkPoolWorker-2] Lock 139657724594272 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-06-19 22:12:04,596: INFO/ForkPoolWorker-2] tde/tde-i18n: rebase remote into repo
[2019-06-19 22:12:11,863: INFO/ForkPoolWorker-2] tdebase/tdmgreet/cs: committing pending changes (addon)
[2019-06-19 22:14:11,880: INFO/ForkPoolWorker-2] Lock 139657724594272 released on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
```

**Comment by nijel** [comment]
I'd be curious to see if Weblate 3.7 makes any difference for you. In case it does not, we should dig deeper into the problems you observe. Just don't forget to restart Celery workers after upgrade ;-).

**Comment by SlavekB** [comment]
After update to 3.7.1, the situation is even worse. No pending translations and commit with POT template update caused nothing – the translation check was not performed, the PO translations was not updated.

```
 * Weblate 3.7.1
 * Python 3.5.3
 * Django 1.11.20
 * Celery 4.2.2
 * celery-batches 0.2
 * six 1.12.0
 * social-auth-core 3.1.0
 * social-auth-app-django 3.1.0
 * django-appconf 1.0.1
 * translate-toolkit 2.4.0
 * translation-finder 1.6
 * Whoosh 2.7.0
 * defusedxml 0.5.0
 * Git 2.11.0
 * Pillow 4.0.0
 * python-dateutil 2.5.3
 * lxml 4.2.5
 * django-crispy-forms 1.6.1
 * django_compressor 2.2
 * djangorestframework 3.9.2
 * user-agents 1.1.0
 * jellyfish 0.7.1
 * diff-match-patch 20121119
 * pytz 2016.7
 * pyuca 1.1.2
 * ruamel.yaml 0.15.97
 * Database backends: django.db.backends.mysql
 * Cache backends: default:RedisCache, avatar:FileBasedCache
 * Email setup: django_sendmail_backend.backends.EmailBackend: localhost
 * Celery: redis://localhost:6379, redis://localhost:6379, regular
 * Platform: Linux 3.16.6-042stab138.42 (x86_64)
```

```
System check identified some issues:

INFOS:
?: (weblate.I021) Error collection is not configured, it is highly recommended for production use
        HINT: https://docs.weblate.org/en/weblate-3.7.1/admin/install.html#collecting-errors

System check identified 1 issue (0 silenced).
```

Related commit: [Gitea – TDE/tde-i18n](https://mirror.git.trinitydesktop.org/gitea/TDE/tde-i18n/commit/439f45c64ce4fcf16210356b01b1f942c6b341a8)
Related component: [Weblate – tdelibs/tdeio](https://mirror.git.trinitydesktop.org/weblate/projects/tdelibs/tdeio/)
Related celery log:
```
[2019-07-04 00:03:30,566: INFO/ForkPoolWorker-4] Lock 140132386979064 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-04 00:03:30,581: INFO/ForkPoolWorker-4] Lock 140132386979064 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-04 00:03:30,785: INFO/ForkPoolWorker-4] tde/tde-i18n: updating repository
[2019-07-04 00:03:30,787: INFO/ForkPoolWorker-4] Lock 140132386979064 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-04 00:03:49,882: INFO/ForkPoolWorker-4] tde/tde-i18n: update took 19.09 seconds
[2019-07-04 00:03:51,728: INFO/ForkPoolWorker-4] Lock 140132386979064 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-04 00:03:56,742: INFO/ForkPoolWorker-4] Lock 140132386979064 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-04 00:03:58,015: INFO/ForkPoolWorker-4] tde/tde-i18n: rebase remote into repo
[2019-07-04 00:06:20,400: INFO/ForkPoolWorker-4] Lock 140132386979064 released on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
```

And after a shorter moment, surprisingly again:
```
[2019-07-04 00:09:39,233: INFO/ForkPoolWorker-4] Lock 140132380641880 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-04 00:09:39,234: INFO/ForkPoolWorker-4] Lock 140132380641880 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-04 00:09:39,350: INFO/ForkPoolWorker-4] tde/tde-i18n: updating repository
[2019-07-04 00:09:39,351: INFO/ForkPoolWorker-4] Lock 140132380641880 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-04 00:09:39,684: INFO/ForkPoolWorker-4] tde/tde-i18n: update took 0.33 seconds
[2019-07-04 00:09:41,381: INFO/ForkPoolWorker-4] Lock 140132380641880 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-04 00:09:43,322: INFO/ForkPoolWorker-4] Lock 140132380641880 released on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
```

However, the corresponding update in Weblate was not performed – the previous string is still provided for translation, the PO files are out of date.

**Comment by nijel** [comment]
Hmm, it turns out that the error would not be logged because of Celery nog logging such errors. Can you please retry after applying 417091902309b28e77e27ce2e4e845b596efa34a?

**Comment by SlavekB** [comment]
Commit 4170919 backported, celery tasks restarted. Today was a situation where the commit was to update the templates and at the same time the translations were waiting to be pushed.

```
==> /var/log/celery/weblate-celery-3.log <==
[2019-07-07 17:10:39,810: INFO/ForkPoolWorker-4] Lock 140134008672608 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-07 17:10:39,811: INFO/ForkPoolWorker-4] Lock 140134008672608 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-07 17:10:39,965: INFO/ForkPoolWorker-4] tde/tde-i18n: updating repository
[2019-07-07 17:10:39,966: INFO/ForkPoolWorker-4] Lock 140134008672608 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-07 17:10:44,316: INFO/ForkPoolWorker-4] tde/tde-i18n: update took 4.35 seconds
[2019-07-07 17:10:45,745: INFO/ForkPoolWorker-4] Lock 140134008672608 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-07 17:10:52,137: INFO/ForkPoolWorker-4] Lock 140134008672608 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-07 17:10:53,301: INFO/ForkPoolWorker-4] tde/tde-i18n: rebase remote into repo

==> /var/log/celery/weblate-notify.log <==
[2019-07-07 17:10:53,334: INFO/MainProcess] Received task: weblate.accounts.tasks.notify_change[b2e0ce3b-9730-46c6-874f-5af7bfe9b6a4]

==> /var/log/celery/weblate-celery-3.log <==
[2019-07-07 17:10:56,809: INFO/ForkPoolWorker-4] tdelibs/tdelibs/cs: committing pending changes (addon)
[2019-07-07 17:12:56,832: INFO/ForkPoolWorker-4] Lock 140134008672608 released on /var/lib/weblate/data/vcs/tde/tde-i18n.lock

==> /var/log/celery/weblate-celery.log <==
[2019-07-07 17:12:56,836: INFO/MainProcess] Received task: weblate.trans.tasks.perform_update[7a21903f-8e8f-41da-9a3e-a7051064f064]  ETA:[2019-07-07 17:13:49.833350+00:00]

==> /var/log/celery/weblate-celery-4.log <==
[2019-07-07 17:13:49,876: INFO/ForkPoolWorker-5] Lock 140134010880408 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-07 17:13:49,878: INFO/ForkPoolWorker-5] Lock 140134010880408 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-07 17:13:50,008: INFO/ForkPoolWorker-5] tde/tde-i18n: updating repository
[2019-07-07 17:13:50,009: INFO/ForkPoolWorker-5] Lock 140134010880408 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-07 17:13:50,263: INFO/ForkPoolWorker-5] tde/tde-i18n: update took 0.25 seconds
[2019-07-07 17:13:51,927: INFO/ForkPoolWorker-5] Lock 140134010880408 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-07 17:13:53,608: INFO/ForkPoolWorker-5] Lock 140134010880408 released on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
```

Nothing was done except for the rebase git repository. Pending translations were not pushed, PO files were not updated, just nothing. There is no error message anywhere. Everything looks happy and fine except that it doesn't do anything. This means that the old strings that have been changed in the templates last week are still offered for translation.

**Comment by nijel** [comment]
Did you restart celery after patching? It seems like there is some problem in committing pending changes as it starts, but doesn't log any commit.

**Comment by SlavekB** [comment]
My previous post started with:
| Commit 4170919 backported, celery tasks restarted.

So yes, celery has been restarted.

Pending commit information was listed in the celery log but no commit was created. By the way, here are more pending translations on linked components.

**Comment by nijel** [comment]
That looks like the error logging is not working as expected, maybe it fails itself.

Anyway let's try to execute that manually to see what is going on, can you please execute following in [management shell](https://docs.weblate.org/en/latest/admin/management.html#invoking-management-commands):

```python
from weblate.trans.models import Component
c = Component.objects.get(slug="tdelibs")
c.commit_pending('test', None)
```

**Comment by SlavekB** [comment]
I see many lines like the following example (email address has been removed), finished information True.

```
INFO tdelibs/tdeio/de: committing pending changes (test)
INFO tdelibs/tdeio/de: committing ['/var/lib/weblate/data/vcs/tde/tde-i18n/tde-i18n-de/messages
/tdelibs/tdeio.po'] as Chris <****>
INFO tdelibs/tdeio_help/de: committing pending changes (test)
INFO tdelibs/tdeio_help/de: committing ['/var/lib/weblate/data/vcs/tde/tde-i18n/tde-i18n-de/mes
sages/tdelibs/tdeio_help.po'] as Chris <****>
…
INFO tdelibs/knotify/de: committing pending changes (test)
INFO tdelibs/knotify/de: committing ['/var/lib/weblate/data/vcs/tde/tde-i18n/tde-i18n-de/messages/tdelibs/knotify.po'] as Chris <****>
True
```

I must point out that celery commit pending changes without problems. For example, after yesterday's (badly successful) update to the GIT repository, 19 commits were pushed. Now this "test" has caused 35 commits to be pushed. The problem is only related to updating the GIT repository from hook.

**Comment by SlavekB** [comment]
By the way, is there a possibility to invoke update the PO files for changed POT's whose changes has been overlooked as a result of the problem?

**Comment by nijel** [comment]
You can also try disabling the webhook temporarily and run the update manually:


```python
from weblate.trans.models import Component
c = Component.objects.get(slug="tdelibs")
c.do_update()
```

The translation files can be updated using [loadpo command](https://docs.weblate.org/en/latest/admin/management.html#loadpo).

**Comment by SlavekB** [comment]
```
INFO tde/tde-i18n: updating repository
INFO tde/tde-i18n: update took 0.80 seconds
True
```
This ignores that the template for several components has been changed and the `weblate.gettext.msgmerge` is not invoked.

If I correctly understand the description of the `loadpo` command, it is used to reload PO files if they have been changed outside the Weblate. Or, it does a templates check and invokes
`weblate.gettext.msgmerge` to update PO files?

**Comment by nijel** [comment]
It only parses the PO files, it indeed does not trigger addon exection.

I think following should trigger the post update addons even if there are no changes:

```python
c.update_branch()
```

**Comment by SlavekB** [comment]
Great, this is exactly what I needed!

By the way, it causes virtually instant commit to all translations that users are doing right now, during `update_branch()`. See current https://mirror.git.trinitydesktop.org/weblate/changes/

**Comment by nijel** [comment]
That's expected - the msgmerge addon forces committing of pending changes so that there are none when updating the PO files.

Great it works this way for you, but what is still to figure out why the automated update doesn't work. I was kind of expecting that this will fail and point us to problem...

**Comment by SlavekB** [comment]
However, here it might be optimized not to commit "all" components but only those for which the POT is updated. The tde-i18n module has 563 pot files... usually not all updated at the same time ;)

Please, must I then run loadpo to update the strings that are updated by msgmerge addon? Although commits have been already pushed, I still see old strings in Weblate.

**Comment by SlavekB** [comment]
Yes, it is "sad" that `update_branch()` was done without any problems, this did not help us in discovering the cause of the problem.

**Comment by nijel** [comment]
Yes, you need loadpo after manually running `update_branch()`, it does not trigger the update.

**Comment by SlavekB** [comment]
By the way, practically regularly after such an update, the Weblate web interface usually ends up on...
```
504 Gateway Timeout

The gateway did not receive a timely response from the upstream server or application.
```
...for some time.

**Comment by nijel** [comment]
It's probably caused by stats update, that should be addressed in future by https://github.com/WeblateOrg/weblate/issues/2786. For now I suggest you increase the timeout on the server (there are also some requests which are expected to take long for now).

**Comment by SlavekB** [comment]
During one recent update, I turned on the DEBUG log for celery:

```
[2019-07-11 00:49:56,726: DEBUG/ForkPoolWorker-5] Attempting to acquire lock 140333438990936 on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:49:56,727: INFO/ForkPoolWorker-5] Lock 140333438990936 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:49:56,727: INFO/ForkPoolWorker-5] Lock 140333438990936 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:49:56,779: INFO/ForkPoolWorker-5] tde/tde-i18n: updating repository
[2019-07-11 00:49:56,780: INFO/ForkPoolWorker-5] Lock 140333438990936 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:50:02,662: INFO/ForkPoolWorker-5] tde/tde-i18n: update took 5.88 seconds
[2019-07-11 00:50:04,907: INFO/ForkPoolWorker-5] Lock 140333438990936 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:50:14,695: INFO/ForkPoolWorker-5] Lock 140333438990936 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:50:15,890: INFO/ForkPoolWorker-5] tde/tde-i18n: rebase remote into repo
[2019-07-11 00:50:18,560: INFO/ForkPoolWorker-5] tdelibs/tdelibs/de: committing pending changes (addon)
[2019-07-11 00:50:18,564: DEBUG/ForkPoolWorker-5] Attempting to acquire lock 140333430165800 on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:50:18,565: DEBUG/ForkPoolWorker-5] Lock 140333430165800 not acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock, waiting 0.05 seconds ...
[2019-07-11 00:50:18,615: DEBUG/ForkPoolWorker-5] Attempting to acquire lock 140333430165800 on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:50:18,616: DEBUG/ForkPoolWorker-5] Lock 140333430165800 not acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock, waiting 0.05 seconds ...
[2019-07-11 00:50:18,666: DEBUG/ForkPoolWorker-5] Attempting to acquire lock 140333430165800 on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:50:18,667: DEBUG/ForkPoolWorker-5] Lock 140333430165800 not acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock, waiting 0.05 seconds ...
⋮
⋮
⋮
[2019-07-11 00:52:18,454: DEBUG/ForkPoolWorker-5] Attempting to acquire lock 140333430165800 on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:52:18,455: DEBUG/ForkPoolWorker-5] Lock 140333430165800 not acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock, waiting 0.05 seconds ...
[2019-07-11 00:52:18,505: DEBUG/ForkPoolWorker-5] Attempting to acquire lock 140333430165800 on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:52:18,506: DEBUG/ForkPoolWorker-5] Lock 140333430165800 not acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock, waiting 0.05 seconds ...
[2019-07-11 00:52:18,556: DEBUG/ForkPoolWorker-5] Attempting to acquire lock 140333430165800 on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:52:18,557: DEBUG/ForkPoolWorker-5] Lock 140333430165800 not acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock, waiting 0.05 seconds ...
[2019-07-11 00:52:18,607: DEBUG/ForkPoolWorker-5] Attempting to acquire lock 140333430165800 on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:52:18,608: DEBUG/ForkPoolWorker-5] Timeout on acquiring lock 140333430165800 on/var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:52:18,608: DEBUG/ForkPoolWorker-5] Attempting to release lock 140333438990936 on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:52:18,609: INFO/ForkPoolWorker-5] Lock 140333438990936 released on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:55:31,069: INFO/ForkPoolWorker-5] Lock 140530381855656 acquired on /var/lib/webl
ate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:55:31,070: INFO/ForkPoolWorker-5] Lock 140530381855656 acquired on /var/lib/webl
ate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:55:31,159: INFO/ForkPoolWorker-5] tde/tde-i18n: updating repository
[2019-07-11 00:55:31,159: INFO/ForkPoolWorker-5] Lock 140530381855656 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:55:31,399: INFO/ForkPoolWorker-5] tde/tde-i18n: update took 0.24 seconds
[2019-07-11 00:55:32,673: INFO/ForkPoolWorker-5] Lock 140530381855656 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:55:34,307: INFO/ForkPoolWorker-5] Lock 140530381855656 released on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:55:38,447: INFO/ForkPoolWorker-5] tde/tde-i18n: pushing to remote repo
[2019-07-11 00:55:38,448: INFO/ForkPoolWorker-5] Lock 140530381855656 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:56:06,817: INFO/ForkPoolWorker-5] Lock 140530381855656 released on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:56:12,515: INFO/ForkPoolWorker-5] Lock 140530381854592 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:56:12,516: INFO/ForkPoolWorker-5] Lock 140530381854592 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:56:12,613: INFO/ForkPoolWorker-5] tde/tde-i18n: updating repository
[2019-07-11 00:56:12,613: INFO/ForkPoolWorker-5] Lock 140530381854592 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:56:12,851: INFO/ForkPoolWorker-5] tde/tde-i18n: update took 0.24 seconds
[2019-07-11 00:56:14,454: INFO/ForkPoolWorker-5] Lock 140530381854592 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:56:15,983: INFO/ForkPoolWorker-5] Lock 140530381854592 released on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:56:19,246: INFO/ForkPoolWorker-5] tde/tde-i18n: pushing to remote repo
[2019-07-11 00:56:19,801: INFO/ForkPoolWorker-5] Lock 140530381854592 acquired on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
[2019-07-11 00:56:24,805: INFO/ForkPoolWorker-5] Lock 140530381854592 released on /var/lib/weblate/data/vcs/tde/tde-i18n.lock
```

The first response to pushing a commit that successfully fetched and rebase repository subsequently failed due to a lock. Where the second response after five minutes delay is taken, I have no idea, but it has not triggered a missed update in Weblate.

Note: I don't see this problem on a repository where there is only one component and no linked components.

**Comment by SlavekB** [comment]
During manual run `c.update_branch()` I sometimes get:

```
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 120, in on_link_wrapper
    return func(self, *args, **kwargs)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 1164, in update_branch
    previous_head=previous_head
  File "/usr/lib/python3/dist-packages/django/dispatch/dispatcher.py", line 193, in send
    for receiver in self._live_receivers(sender)
  File "/usr/lib/python3/dist-packages/django/dispatch/dispatcher.py", line 193, in <listcomp>
    for receiver in self._live_receivers(sender)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/addons/models.py", line 186, in post_update
    addon.addon.post_update(component, previous_head)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/addons/base.py", line 310, in post_update
    self.commit_and_push(component)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/addons/base.py", line 295, in commit_and_push
    with repository.lock:
  File "/var/lib/weblate/lib/python3.5/site-packages/filelock.py", line 323, in __enter__
    self.acquire()
  File "/var/lib/weblate/lib/python3.5/site-packages/filelock.py", line 278, in acquire
    raise Timeout(self._lock_file)
filelock.Timeout: The file lock '/var/lib/weblate/data/vcs/tde/tde-i18n.lock' could not be acquired.
```

**Comment by SlavekB** [comment]
In today's regular update, I observed that the same problem occurred also with another repository that contained a significantly smaller number of linked components – 29 components.

**Comment by paulpach** [comment]
Ok,  there are at least 2 issues here.  

For some reason [this](https://github.com/WeblateOrg/weblate/blob/2b725379651f749bc0b291804612aafe51015685/weblate/addons/base.py#L207) catch is not working as expected.  It is never catching the CalledProcessError,   I added another catch like this:

```py
def execute_process(self, component, cmd, env=None):
        component.log_debug('%s addon exec: %s', self.name, repr(cmd))
        try:
            output = subprocess.check_output(
                cmd,
                env=get_clean_env(env),
                cwd=component.full_path,
                stderr=subprocess.STDOUT,
            )
            component.log_debug('exec result: %s', repr(output))
        except (OSError, subprocess.CalledProcessError) as err:
            output = getattr(err, 'output', '').decode('utf-8')
            component.log_error('failed to exec %s: %s', repr(cmd), err)
            for line in output.splitlines():
                component.log_error('program output: %s', line)
            self.alerts.append({
                'addon': self.name,
                'command': ' '.join(cmd),
                'output': output,
                'error': str(err),
            })
        except Exception as ex:
            component.log_error('failed to exec %s: %s', ex.cmd, ex.output)
```
And now I can see the error.   Without this extra except,  for me it tries to run the first language,  then fails,  the exception is never caught and it is silently discarded.  The other languages are never run.

to test this,  add a duplicate string in your pot.   You will see it try to run the first language and then nothing else,  no error no nothing.


there is another issue,  which is that it is not calling the addon sometimes at all.   I am troubleshooting that one.

**Comment by ss-gxp** [comment]
Not sure if this is related, but it looks like it.
Version 3.7.1 in docker. The project with the base component (updated by the gitlab hook) and more than 80 components linked with it. The celery task is restarted and successfully runs, but the changes are not imported. To correct the situation, I perform reset, but it happens almost every time.

I did not look deeply, but a possible problem is that it is not possible to acquire lock in the linked components, since acquired lock base component? And when the task is restarted, changes in the repository are not detected?

All these assumptions, of course. But maybe the traceback (from Flower UI) will help find the answer:

> Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/celery/app/base.py", line 472, in run
    return task._orig_run(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/weblate/trans/tasks.py", line 63, in perform_update
    obj.do_update()
  File "/usr/local/lib/python3.7/dist-packages/weblate/trans/models/component.py", line 120, in on_link_wrapper
    return func(self, *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/weblate/trans/models/component.py", line 862, in do_update
    ret = self.update_branch(request, method=method)
  File "/usr/local/lib/python3.7/dist-packages/weblate/trans/models/component.py", line 120, in on_link_wrapper
    return func(self, *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/weblate/trans/models/component.py", line 1164, in update_branch
    previous_head=previous_head
  File "/usr/local/lib/python3.7/dist-packages/django/dispatch/dispatcher.py", line 175, in send
    for receiver in self._live_receivers(sender)
  File "/usr/local/lib/python3.7/dist-packages/django/dispatch/dispatcher.py", line 175, in <listcomp>
    for receiver in self._live_receivers(sender)
  File "/usr/local/lib/python3.7/dist-packages/weblate/addons/models.py", line 186, in post_update
    addon.addon.post_update(component, previous_head)
  File "/usr/local/lib/python3.7/dist-packages/weblate/addons/base.py", line 310, in post_update
    self.commit_and_push(component)
  File "/usr/local/lib/python3.7/dist-packages/weblate/addons/base.py", line 295, in commit_and_push
    with repository.lock:
  File "/usr/local/lib/python3.7/dist-packages/filelock.py", line 323, in __enter__
    self.acquire()
  File "/usr/local/lib/python3.7/dist-packages/filelock.py", line 278, in acquire
    raise Timeout(self._lock_file)
filelock.Timeout: The file lock '/app/data/vcs/freedom/base.lock' could not be acquired.
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/celery/app/trace.py", line 385, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/celery/app/trace.py", line 648, in __protected_call__
    return self.run(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/celery/app/base.py", line 481, in run
    raise task.retry(exc=exc, **retry_kwargs)
  File "/usr/local/lib/python3.7/dist-packages/celery/app/task.py", line 725, in retry
    raise ret
celery.exceptions.Retry: Retry in 347s: Timeout()

**Comment by nijel** [comment]
Well the tasks are supposed to retry on failed lock since ec340305d75 and you seem to get the Retry exception what is correct way of doing that in Celery. The task state should be RETRY at that point.

**Comment by ss-gxp** [comment]
Yes, that is right. And after the restart, it ends with the status "SUCCESS", but does not see any changes to the source lines in the repository.

**Comment by SlavekB** [comment]
I tried to use the code from the [yesterday's comment](issues/2813#issuecomment-512649231) but nothing changed – no error message, no updates. I again had to manually update.

I note that the problem occurs regardless of whether there are pending translations to commit here.

**Comment by nijel** [comment]
Looking again at the traceback in https://github.com/WeblateOrg/weblate/issues/2813#issuecomment-513235804, I think were problem might be (it's clearly a bug, but it might not be the same bug that all you are hitting). The lock is released after update and the post update tries to acquire then again. When it fails (eg. there is another update in the queue and it takes it), it will not properly recover as the update is already stored in the database and the post update signal won't be sent again. This should be fixed by 19a75412fdf2e6b1d9f3c6d96bb131a4feec20be.

**Comment by ss-gxp** [comment]
I tried to use the patch 19a7541 (applying to 3.7.1) - linked components is not updated, only the base component is being updated. Maybe I missed something.

**Comment by nijel** [comment]
Do you get some error in the logs? You might need to apply https://github.com/WeblateOrg/weblate/commit/417091902309b28e77e27ce2e4e845b596efa34a to see errors in the Celery logs.

**Comment by SlavekB** [comment]
I have backported 4170919, 6859e3e, 19a7541 and code from [#2813 comment](2813#issuecomment-512649231) and the problem persists.

**Comment by SlavekB** [comment]
So I have to make it clear that the situation is even worse now. When I proceeded to manually start the update from Weblate shell, as in the past few weeks successfully:
```
from weblate.trans.models import Component
c = Component.objects.get(slug="tde-i18n")
c.update_branch()
```

This time it failed because of the lock:
```
INFO tde/tde-i18n: rebase remote into repo
INFO tdebase/kappfinder/de: committing pending changes (addon)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 120, in on_link_wrapper
    return func(self, *args, **kwargs)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 1165, in update_branch
    child=True,
  File "/usr/lib/python3/dist-packages/django/dispatch/dispatcher.py", line 193, in send
    for receiver in self._live_receivers(sender)
  File "/usr/lib/python3/dist-packages/django/dispatch/dispatcher.py", line 193, in <listcomp>
    for receiver in self._live_receivers(sender)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/addons/models.py", line 188, in post_update
    addon.addon.post_update(component, previous_head)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/addons/base.py", line 310, in post_update
    component.commit_pending('addon', None, skip_push=True)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 118, in on_link_wrapper
    *args, **kwargs
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 120, in on_link_wrapper
    return func(self, *args, **kwargs)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 120, in on_link_wrapper
    return func(self, *args, **kwargs)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 1053, in commit_pending
    reason, request, skip_push=True, force=True
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/translation.py", line 433, in commit_pending
    with self.component.repository.lock:
  File "/var/lib/weblate/lib/python3.5/site-packages/filelock.py", line 323, in __enter__
    self.acquire()
  File "/var/lib/weblate/lib/python3.5/site-packages/filelock.py", line 278, in acquire
    raise Timeout(self._lock_file)
filelock.Timeout: The file lock '/var/lib/weblate/data/vcs/tde/tde-i18n.lock' could not be acquired.
```

So now both automatic and manual updates fail.

**Comment by nijel** [comment]
There is some other update running holding the lock?

**Comment by nijel** [comment]
Actually it might be caused by the same update, that should be addressed by da3c3430ecd67d5e3fe01482d293f286a25e19b4.

**Comment by SlavekB** [comment]
You may notice that there was a proper rebase first (although there was nothing to do at the time) and the immediate step in the same update failed. So there was no other update here == very likely this is a failure caused by the same update.

Once the update that I have now run manually is complete, I apply the above patches again, along with da3c343.

**Comment by SlavekB** [comment]
Here is one step forward when I backported 4170919, 6859e3e, 19a7541 and da3c343: If there are no pending changes in the repository, the branch update will succeed.

If there are pending changes, the commit attempt fails due to the lock:

```
(InteractiveConsole)
>>> from weblate.trans.models import Component
>>> c = Component.objects.get(slug="tde-i18n")
>>> c.update_branch()
INFO tde/tde-i18n: rebase remote into repo
INFO tdelibs/tdelibs/cs: committing pending changes (addon)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 120, in on_link_wrapper
    return func(self, *args, **kwargs)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 1168, in update_branch
    child=True,
  File "/usr/lib/python3/dist-packages/django/dispatch/dispatcher.py", line 193, in send
    for receiver in self._live_receivers(sender)
  File "/usr/lib/python3/dist-packages/django/dispatch/dispatcher.py", line 193, in <listcomp>
    for receiver in self._live_receivers(sender)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/addons/models.py", line 188, in post_update
    addon.addon.post_update(component, previous_head)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/addons/base.py", line 310, in post_update
    component.commit_pending('addon', None, skip_push=True)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 118, in on_link_wrapper
    *args, **kwargs
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 120, in on_link_wrapper
    return func(self, *args, **kwargs)
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/component.py", line 1056, in commit_pending
    reason, request, skip_push=True, force=True
  File "/var/lib/weblate/lib/python3.5/site-packages/weblate/trans/models/translation.py", line 433, in commit_pending
    with self.component.repository.lock:
  File "/var/lib/weblate/lib/python3.5/site-packages/filelock.py", line 323, in __enter__
    self.acquire()
  File "/var/lib/weblate/lib/python3.5/site-packages/filelock.py", line 278, in acquire
    raise Timeout(self._lock_file)
filelock.Timeout: The file lock '/var/lib/weblate/data/vcs/tde/tde-i18n.lock' could not be acquired.
>>>
```

If I push changes before `update_branch` (using `weblate commitgit tde/tde-i18n`) and at the same time there are no changes during `update_branch`, changes from POT files are successfully merged into PO files and these are pushed into GIT.

**Comment by nijel** [comment]
Thanks for testing that! With trackeback it's much easier to figure out what went wrong :-). This should be addressed by 9a4b0c1b3d.

**Comment by SlavekB** [comment]
Great, now it looks good. Backported 4170919, 6859e3e, 19a7541, da3c343, 9a4b0c1 and e495700.

After commit to git repository, commits for pending translations are created, commit is made in connection with POT template update, "loadpo" is performed. And during "loadpo" processing, the commits were pushed into the git repository.

Everything now seems to be as it should be. Thank you!

**Comment by nijel** [comment]
Thanks a lot for testing. I'm closing the issue for now.

* In case you see problem with the fix, please comment on this issue. 
* In case you see similar problem, please open separate issue.
* If you are happy with the outcome, consider supporting Weblate by [donating](https://weblate.org/donate/).
